solrmonitor/SolrEventListener.go

```diff
@@ -0,0 +1,38 @@
+package solrmonitor
+
+/**
+	The client can register the SolrEventListener to listen to the solr cluster state events from the zookeeper.
+
+	On startup, the events are fired in this order:
+		1. SolrLiveNodesChanged fires for all the solr live nodes (node which keeps indexes). It keeps watch on zk node /solr/live_nodes 's children
+		2. SolrQueryNodesChanged fires for all the solr query nodes (node which coordinates the query execution). It keeps watch on zk node /solr/live_query_nodes 's children
+		3. SolrCollectionsChanged fires for all collections that exist. It keeps watch on zk node /solr/collections 's children.
+			3.1 SolrCollectionStateChanged fires for each collection; keeps watch on zk node /solr/collections/collname/state.json
+			3.2 SolrCollectionReplicaStatesChanged fires if Per replica state(PRS) is enable for collection.
+				If PRS enable then it keeps watch on zk node /solr/collections/collname/state.json 's children. Those children represents state of each replica of collection
+				(reference https://issues.apache.org/jira/browse/SOLR-15052, https://docs.google.com/document/d/1xdxpzUNmTZbk0vTMZqfen9R3ArdHokLITdiISBxCFUg/edit )
+
+	After initialization it delivers events for following condition
+		1. SolrLiveNodesChanged if solr live nodes changes
+		2. SolrQueryNodesChanged if solr query node changes
+		3. SolrCollectionsChanged if number of collections changes in solr cluster
+		4. SolrCollectionStateChanged if collection's number of shards changes, or replica moves, or replica splits.
+		5. SolrCollectionReplicaStatesChanged if collection's replica goes up/down, or becomes leader
```

Comment:
only for collections that operate under PRS?

Reply:
Yeah. we will have two modes. Non-prs collection will keep all collection-state into state.json file and they will update replica up/down state in state.json file. But, prs collection will update replica up/down state separately(i.e. they will not update state.json file), but state.json file will be updated if #shards changes, replica moves, or shard split.